ord_schema/dataset_templating_test.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2020 Open Reaction Database Project Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for ord_schema.message_helpers."""
+
+import pandas as pd
+
+from absl.testing import absltest
+from google.protobuf import text_format
+
+from ord_schema import dataset_templating
+from ord_schema.proto import reaction_pb2
+from ord_schema.proto import dataset_pb2
+
+
+class DatasetTemplatingTest(absltest.TestCase):
+    def setUp(self):
+        super().setUp()
+
+        message = reaction_pb2.Reaction()
+        dummy_input = message.inputs['in']
+        outcome = message.outcomes.add()
+        component = dummy_input.components.add()
+        component.identifiers.add(type='SMILES', value='CCO')
+        component.is_limiting = True
+        component.mass.value = 1
+        component.mass.units = reaction_pb2.Mass.GRAM
+        outcome.conversion.value = 75
+        self.valid_reaction = message
+
+    def test_valid_templating(self):
+        template_string = text_format.MessageToString(self.valid_reaction)
+        template_string = template_string.replace('value: "CCO"',
+                                                  'value: "$my_smiles$"')
+        template_string = template_string.replace('value: 75',
+                                                  'value: $conversion$')
+        df = pd.DataFrame.from_dict({
+            '$my_smiles$': ['CCO', 'CCCO', 'CCCCO'],
+            '$conversion$': ['75', '50', '30'],
+        })
+        dataset = dataset_templating.generate_dataset(template_string, df)
+        exptected_reactions = []
+        for smiles, conversion in zip(['CCO', 'CCCO', 'CCCCO'], [75, 50, 30]):
+            reaction = reaction_pb2.Reaction()
+            reaction.CopyFrom(self.valid_reaction)
+            reaction.inputs['in'].components[0].identifiers[0].value = smiles
+            reaction.outcomes[0].conversion.value = conversion
+            exptected_reactions.append(reaction)
+        expected_dataset = dataset_pb2.Dataset(reactions=exptected_reactions)
+        self.assertEqual(dataset, expected_dataset)
+
+        df = pd.DataFrame.from_dict({
```

Comment:
duplicated code from above?

Reply:
Slightly different: column name is `my_smiles` instead of `$my_smiles$`, for example

Reply:
Ah; please add a comment so that's clear to readers.